packages/nimble-components/src/dialog/index.ts

```diff
@@ -125,11 +124,33 @@ export class Dialog<CloseReason = void> extends FoundationElement {
         if (this.preventDismiss) {
             event.preventDefault();
         } else {
-            this.resolveShow!(UserDismissed);
-            this.resolveShow = undefined;
+            this.notifyClosed(UserDismissed);
         }
         return true;
     }
+
+    /**
+     * @internal
+     */
+    public closeHandler(): void {
+        if (this.resolveShow) {
+            // If
+            // - the browser implements dialogs with the CloseWatcher API, and
+            // - the user presses ESC without first interacting with the dialog (e.g. clicking, scrolling),
+            // the cancel event is not fired and the dialog just closes.
+            this.notifyClosed(UserDismissed);
+        }
+    }
+
+    private notifyClosed(reason: CloseReason | UserDismissed): void {
+        if (!this.resolveShow) {
+            throw new Error(
+                'Do not call notifyClosed unless there is a promise to resolve'
```

Comment:
Can a user end up with this error being called or is this an internal error / guard against implementation errors?

If it's user facing / possible for a user to trigger the code path with the public api can we have a test to cover it and frame the error for the user?

Reply:
> or is this an internal error / guard against implementation errors?

Yes, that.